blocks/datasets/__init__.py

```diff
@@ -0,0 +1,233 @@
+
+        """
+        pass
+
+    def close(self, state):
+        """Cleanly close the dataset e.g. close file handles."""
+        pass
+
+    @abstractmethod
+    def get_data(self, state=None, request=None, sources=None):
+        """Request data from the dataset.
+
+        Parameters
+        ----------
+        state : object, optional
+            The state as returned by the :meth:`open` method. The dataset
+            can use this to e.g. interact with files when needed.
+        request : object, optional
+            If supported, the request for a particular part of the data
+            e.g. the number of examples to return, or the indices of a
+            particular minimbatch of examples.
+        sources : tuple of strings, optional
+            The data sources to return. By default all data sources are
+            returned.
+
+        .. todo::
+
+           A way for the dataset to communicate which kind of requests it
+           accepts, and a way to communicate what kind of request is being
+           sent when supporting multiple.
+
+        Returns
+        -------
+        tuple
+            A tuple of data.
+
+        """
+        raise NotImplementedError
+
+
+class DataStream(object):
+    """A stream of data.
+
+    A data stream implements the ``__iter__`` protocol, returning a new
+    iterator over the data each time it gets called (in the case of finite
+    data). A single pass over one of these iterators represents an *epoch*.
+
+    The data stream also opens a given dataset and maintains the state
+    (e.g. file handles) if necessary.
+
+    A data stream can provide an iterator over a dataset, but it can also
+    wrap another datastream e.g. to provide caching or some sort of
+    pre-processing.
+
+    Parameters
+    ----------
+    data : :class:`Dataset` or :class:`DataStream`
+        An object that implements the :meth:`open` and :meth:`get_data`
+        methods. Usually this is a :class:`Dataset` instance, but when
+        chaining iterators to e.g. perform caching this can also be a
+        :class:`DataStream` object.
+    iteration_scheme : :class:`IterationScheme`, optional
+        The iteration scheme to use when retrieving data. Note that not all
+        datasets support the same iteration schemes, some data sets require
+        one, and others don't support any.
+    sources : tuple of strings, optional
+        The sources of data to return. By default, all sources of the
+        dataset are requested.
+
+    """
+    def __init__(self, data, iteration_scheme=None, sources=None):
+        update_instance(self, locals())
+        self.data_state = None
+
+    def __iter__(self):
+        self.data_state = self.data.open(self.data_state)
+        # TODO Allow for multiple child datasets/data streams, so that they
+        # can be combined
+        if isinstance(self.data, DataStream):
+            self.child_iterator = iter(self.data)
```

Comment:
It should only be called once right now? I'll double check the code when I
can, but it should only be called by the __iter__ of the data stream, once
before each epoch.
On Jan 9, 2015 9:43 AM, "Dmitry Bogdanov" <notifications@github.com> wrote:

> In blocks/datasets/__init__.py
> <https://github.com/bartvm/blocks/pull/91#discussion-diff-22705740>:
>
> > +        one, and others don't support any.
> > +    sources : tuple of strings, optional
> > +        The sources of data to return. By default, all sources of the
> > +        dataset are requested.
> > +
> > +    """
> > +    def __init__(self, data, iteration_scheme=None, sources=None):
> > +        update_instance(self, locals())
> > +        self.data_state = None
> > +
> > +    def __iter__(self):
> > +        self.data_state = self.data.open(self.data_state)
> > +        # TODO Allow for multiple child datasets/data streams, so that they
> > +        # can be combined
> > +        if isinstance(self.data, DataStream):
> > +            self.child_iterator = iter(self.data)
>
> Doesn't iter call self.data.open under the hood? It is not nice that it
> is called twice.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bartvm/blocks/pull/91/files#r22705740>.
>